homeassistant/components/hive/__init__.py

```diff
@@ -69,75 +64,113 @@
+                    CONF_PASSWORD: conf[CONF_PASSWORD],
+                    CONF_SCAN_INTERVAL: conf.get(CONF_SCAN_INTERVAL, 120),
+                },
+            )
+        )
+    return True
+
+
+async def async_setup_entry(hass, entry):
+    """Set up Hive from a config entry."""
+    # Store an API object for your platforms to access
+    # hass.data[DOMAIN][entry.entry_id] = MyApi(...)
+
+    websession = aiohttp_client.async_get_clientsession(hass)
+    hive = Hive(websession)
+    hive_config = dict(entry.data)
+    hive_options = dict(entry.options)
 
     async def heating_boost(service):
         """Handle the service call."""
 
         entity_lookup = hass.data[DOMAIN]["entity_lookup"]
-        hive_id = entity_lookup.get(service.data[ATTR_ENTITY_ID])
-        if not hive_id:
+        device = entity_lookup.get(service.data[ATTR_ENTITY_ID])
+        if not device:
             # log or raise error
             _LOGGER.error("Cannot boost entity id entered")
             return
 
         minutes = service.data[ATTR_TIME_PERIOD]
         temperature = service.data[ATTR_TEMPERATURE]
 
-        hive.heating.turn_boost_on(hive_id, minutes, temperature)
+        await hive.heating.turn_boost_on(device, minutes, temperature)
 
     async def hot_water_boost(service):
         """Handle the service call."""
         entity_lookup = hass.data[DOMAIN]["entity_lookup"]
-        hive_id = entity_lookup.get(service.data[ATTR_ENTITY_ID])
-        if not hive_id:
+        device = entity_lookup.get(service.data[ATTR_ENTITY_ID])
+        if not device:
             # log or raise error
             _LOGGER.error("Cannot boost entity id entered")
             return
+
         minutes = service.data[ATTR_TIME_PERIOD]
-        mode = service.data[ATTR_MODE]
+        mode = service.data[ATTR_ONOFF]
 
         if mode == "on":
-            hive.hotwater.turn_boost_on(hive_id, minutes)
+            await hive.hotwater.turn_boost_on(device, minutes)
         elif mode == "off":
-            hive.hotwater.turn_boost_off(hive_id)
+            await hive.hotwater.turn_boost_off(device)
 
-    hive = Hive()
+        # Update config entry options
 
-    config = {}
-    config["username"] = config[DOMAIN][CONF_USERNAME]
-    config["password"] = config[DOMAIN][CONF_PASSWORD]
-    config["update_interval"] = config[DOMAIN][CONF_SCAN_INTERVAL]
-
-    devices = await hive.session.startSession(config)
+    hive_options = hive_options if len(hive_options) > 0 else hive_config["options"]
+    hive_config["options"].update(hive_options)
+    hive_config["add_sensors"] = bool("hive" in hass.data["custom_components"])
+    hass.config_entries.async_update_entry(entry, options=hive_options)
+    hass.data[DOMAIN][entry.entry_id] = hive
+    hass.data[DOMAIN]["entity_lookup"] = {}
 
-    if devices is None:
-        _LOGGER.error("Hive API initialization failed")
-        return False
+    try:
+        hive.devices = await hive.session.startSession(hive_config)
```

Comment:
Why are we modifying the `Hive` instance? I'd expect the `Hive` class to be responsible for that.